### PR TITLE
remove slonik, insert postgres.js

### PR DESCRIPTION
**Pull request by issa-tseng**

### CODE REVIEW

**Comment by matthew-white** [comment]
I'm surprised it's not a larger diff! It's nice that there are a few similarities between postgres.js and Slonik.

**Review by ktuite — APPROVED** [review]

**Comment by matthew-white** [comment]
I just followed the repro steps in #482 in order to see whether #604 fixed that issue. However, I forgot to checkout #604 first, so I ended up running the repro steps against this PR. 😅 The result was interesting though, so I thought I'd note what I saw.

In good news, I'm no longer observing a database connection leak. 🎉 After I followed the repro steps, the `pg_stat_activity` count did not increase.

I ran the repro steps several times. Every time that I checked the number of successful requests, 8/10 requests were successful. That's interesting to me, because with Slonik today, it's 9/10. From a quick look, it seems like postgres.js defaults to allowing up to 10 database connections, the same as Slonik, so that doesn't seem to explain the difference.

I queried `pg_stat_activity` during the repro steps, running `select count(*) from pg_stat_activity where state <> 'idle'`. From what I can tell, the count spikes to 9, quickly decreases to 8, then continues to decrease. I'm surprised not to see a count of more like 10 or 11, which is what I'd expect if postgres.js were using 10 connections.

The requests that don't succeed return with a variety of errors. Usually it's a 500 or a 504, but I've also seen some 502s. When it's a 500, Backend is returning an error response whose `message` property starts with "Completely unhandled exception". All the 504s and 502s I checked were HTML responses from nginx.

The specific error details also varied. For 500s, those details showed up in the error response. In other cases, they showed up in Sentry. The two most common errors were:

1. portal "" does not exist (Sentry [[1]](https://sentry.io/organizations/getodk/issues/3584049649/)[[2]](https://sentry.io/organizations/getodk/issues/3584122643/)[[3]](https://sentry.io/organizations/getodk/issues/3584149324/))
2. RangeError: The value of "offset" is out of range. &hellip; (Sentry [[1]](https://sentry.io/organizations/getodk/issues/3584049670/)[[2]](https://sentry.io/organizations/getodk/issues/3584149405/))

I also saw these 1 or 2 times:

3. TypeError: query.cursorFn is not a function ([Sentry](https://sentry.io/organizations/getodk/issues/3584176797/))
4. TypeError: constructors[k] is not a function ([Sentry](https://sentry.io/organizations/getodk/issues/3584190574/))
5. SyntaxError: Unexpected number in JSON at position 4 ([Sentry](https://sentry.io/organizations/getodk/issues/3584269382/))

(5) happened when I tried the repro steps with 20 concurrent requests instead of 10. I'm pretty sure that's true of (3) as well, and that might also be true of (4). Interestingly, when I tried 20 requests, there was also one time when I received a 404.1 Problem response ("Could not find the resource you were looking for."). There were also times when `curl` exited with an exit code of 6, displaying the message "Could not resolve host: test.getodk.cloud". That said, sending 20 concurrent requests resulted in more successful responses than sending 10 requests. (The count of successful requests seemed to vary. I saw 10 once.)

I have a theory about when a 500 is returned vs. a 504. Errors (1) and (2) above were thrown for `stream`, `maybeOne`, and `oneFirst`. I think I'm seeing a pattern that when `maybeOne` and `oneFirst` throw the error, it's a 500, while when it's `stream`, it's a 504.

**Comment by matthew-white** [comment]
@lognaturel also noted on Slack:

> For what it's worth, I was disrupted at least once right when Sentry got that portal error. I was attempting to publish a form draft in one tab and had another form’s submission page open in another. The publish request seemed to hang. I got a spinner on the modal. I got impatient and killed the window. When I came back the publish had succeeded.

I think this would be expected if the publish request was sent while all available database connections were in use by the stream requests. With Slonik, a connection attempt times out in 5 seconds, so you wouldn't expect the request to hang, but with postgres.js, it looks like the timeout is 30 seconds.

File: lib/external/postgres.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    // we don't want to automatically assume all non-true values can be safely
+    // equal to 'f', since we can take user input here.
+    boolean: {
+      to: 16, from: 16,
+      serialize: (x => (x === true ? 't' : x === false ? 'f' : x)),
+      parse: (x => x === 't')
+    }
+  }
+};
+const postgres = (config) => _postgres(connectionString(config), options);
```
[the remark this answers is not in the document]
Comment (matthew-white) [reply lib/external/postgres.js:35]: #433 is about Slonik, but I think has some relevant discussion.

**Comment by matthew-white** [comment]
postgres.js branch that we could try to see if it fixes the issues we were seeing: https://github.com/porsager/postgres/tree/concurrent-cursors

**Comment by issa-tseng** [comment]
no, it doesn’t, don’t bother, it won’t. this is shared as a possible starting point for us implementing this ourselves. On Aug 15, 2023, at 16:04, Matthew White ***@***.***> wrote:﻿
postgres.js branch that we could try to see if it fixes the issues we were seeing: https://github.com/porsager/postgres/tree/concurrent-cursors

**Comment by lognaturel** [comment]
Closing in favor of tracking issue at https://github.com/getodk/central/issues/614

**Pull request closed without merging**